doc/source/data/advanced-pipelines.rst

```diff
@@ -175,7 +175,7 @@ For example, in the following pipeline, the ``map(func)`` transformation only oc
 
 .. important::
 
-    Result caching only applies if there are *transformation* stages prior to the pipelining operation. If you ``repeat()`` or ``window()`` a Dataset right after the read call (e.g., ``ray.data.read_parquet(...).repeat()``), then the read will still be re-executed on each repetition. This optimization saves memory, at the cost of repeated reads from the datasource.
+    Result caching only applies if there are *transformation* stages prior to the pipelining operation. If you ``repeat()`` or ``window()`` a Dataset right after the read call (e.g., ``ray.data.read_parquet(...).repeat()``), then the read will still be re-executed on each repetition. This optimization saves memory, at the cost of repeated reads from the datasource. To force result caching in all cases, use ``.fully_executed().repeat()``.
```

Comment:
I feel we have a bit too many gotchas / if-else cases and it may hurt understandability/usability.  Since we rely on .fully_execute(), do you think we can simplify it as "it will *always* re-execute starting from the node with .fully_execute()"?

Reply:
Unfortunately, not without breaking eager semantics. Agree it's confusing but maybe we can compensate with over-explaining a bit here.